File: backend/product-service/src/entities/base.entity.go
package entities

import (
	"time"

	"gorm.io/gorm"
)

// BaseEntity contains common fields for all entities
type BaseEntity struct {
	CreatedAt time.Time      `gorm:"not null;default:CURRENT_TIMESTAMP;index" json:"created_at"`
	UpdatedAt time.Time      `gorm:"not null;default:CURRENT_TIMESTAMP" json:"updated_at"`
	DeletedAt gorm.DeletedAt `gorm:"index" json:"deleted_at,omitempty"`
	
	// Audit fields
	CreatedBy *string `gorm:"type:varchar(255)" json:"created_by,omitempty"`
	UpdatedBy *string `gorm:"type:varchar(255)" json:"updated_by,omitempty"`
	DeletedBy *string `gorm:"type:varchar(255)" json:"deleted_by,omitempty"`
	
	// Version for optimistic locking
	Version int `gorm:"default:1" json:"version"`
}

// BeforeCreate hook for BaseEntity
func (base *BaseEntity) BeforeCreate(tx *gorm.DB) error {
	now := time.Now()
	base.CreatedAt = now
	base.UpdatedAt = now
	base.Version = 1
	return nil
}

// BeforeUpdate hook for BaseEntity
func (base *BaseEntity) BeforeUpdate(tx *gorm.DB) error {
	base.UpdatedAt = time.Now()
	base.Version++
	return nil
}

// SoftDelete performs a soft delete
func (base *BaseEntity) SoftDelete(deletedBy *string) {
	now := time.Now()
	base.DeletedAt = gorm.DeletedAt{Time: now, Valid: true}
	base.DeletedBy = deletedBy
	base.UpdatedAt = now
}

// IsDeleted checks if the entity is soft deleted
func (base *BaseEntity) IsDeleted() bool {
	return base.DeletedAt.Valid
}

// Restore restores a soft deleted entity
func (base *BaseEntity) Restore() {
	base.DeletedAt = gorm.DeletedAt{}
	base.DeletedBy = nil
	base.UpdatedAt = time.Now()
}

// SetAuditInfo sets audit information
func (base *BaseEntity) SetAuditInfo(createdBy, updatedBy *string) {
	if base.CreatedBy == nil {
		base.CreatedBy = createdBy
	}
	base.UpdatedBy = updatedBy
}

File: backend/user-service/src/Domain/Entities/BaseEntity.cs
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;

namespace UserService.Domain.Entities;

public abstract class BaseEntity
{
    [Column("created_at")]
    public DateTime CreatedAt { get; set; } = DateTime.UtcNow;

    [Column("updated_at")]
    public DateTime UpdatedAt { get; set; } = DateTime.UtcNow;

    [Column("created_by")]
    [MaxLength(255)]
    public string? CreatedBy { get; set; }

    [Column("updated_by")]
    [MaxLength(255)]
    public string? UpdatedBy { get; set; }

    [Column("version")]
    [Timestamp]
    public byte[]? Version { get; set; }

    [Column("is_deleted")]
    public bool IsDeleted { get; set; } = false;

    [Column("deleted_at")]
    public DateTime? DeletedAt { get; set; }

    [Column("deleted_by")]
    [MaxLength(255)]
    public string? DeletedBy { get; set; }

    public void SoftDelete(string? deletedBy = null)
    {
        IsDeleted = true;
        DeletedAt = DateTime.UtcNow;
        DeletedBy = deletedBy;
    }

    public void Restore()
    {
        IsDeleted = false;
        DeletedAt = null;
        DeletedBy = null;
    }

    public void UpdateTimestamp(string? updatedBy = null)
    {
        UpdatedAt = DateTime.UtcNow;
        UpdatedBy = updatedBy;
    }
}
